Optionally enable caching of any request method

tuupola opened the pull request:
| Q               | A
| --------------- | ---
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| Deprecations?   | no
| Documentation   | https://github.com/php-http/documentation/pull/174
| License         | MIT

#### What's in this PR?

Adds  new `methods` setting which allows you to change which request methods are allowed to be cached.

#### Why?

Sometimes it is desirable to cache `POST` requests. For example when using HTTPlug based SOAP client. I am currently using this in live system for SOAP requests.

#### Example Usage

To enable caching of `GET` and `POST` requests you can do the following.

```php
$cache = new CachePlugin($pool, $streamFactory, [
    "default_ttl" => 3600 * 48,
    "respect_cache_headers" => true,
    "methods" => ["GET", "POST"]
]);
```

#### Checklist

- [x] Updated CHANGELOG.md to describe new feature.
- [x] Documentation pull request created.

#### To Do

- [x] Should PUT and PATCH et al. also be allowed if user requests so?
- [x] Make sure tests are actually ok.

tuupola committed:
Enable caching of POST requests when requested

This is useful for example when using a HTTPlug based SoapClient.
php-http/client-common#26

tuupola committed:
Make StyleCI happy

dbu commented on src/CachePlugin.php:
```diff
@@ -273,12 +276,17 @@ private function configureOptions(OptionsResolver $resolver)
+        $resolver->setAllowedTypes('methods', 'array');
         $resolver->setAllowedValues('hash_algo', hash_algos());
+        $resolver->setAllowedValues('methods', function ($value) {
+            return 0 === count(array_diff($value, ['GET', 'HEAD', 'POST']));
```
i think once we allow to configure, we should allow any methods. we need a big warning sign for this anyways, as people should know what they do before caching post and other potentially modifying requests... for example, it could make sense to cache a PROPFIND request (webdav, its a non-modifying request)...

tuupola replied:
I agree. Question is since RFC2616 says valid request method can be pretty much any string (not sure if A..Z or alphanumeric) should any request method be allowed or just the common methods defined in [RFC2616-sec9](https://www.w3.org/Protocols/rfc2616/rfc2616-sec9.html)

dbu replied:
i'd not do any validation apart from requiring an array of strings.

tuupola replied:
Found the spec. RFC2616 says request method should be a token and [RFC7230](https://tools.ietf.org/html/rfc7230#section-3.2.6) defines token to be any visible `USASCII` character except delimiters `"(),/:;<=>?@[\]{}`.

dbu commented on src/CachePlugin.php:
```diff
@@ -225,6 +224,10 @@ private function getCacheControlDirective(ResponseInterface $response, $name)
     private function createCacheKey(RequestInterface $request)
     {
+        if ('POST' === $request->getMethod()) {
+            return hash($this->config['hash_algo'], $request->getMethod().' '.$request->getUri().' '.$request->getBody());
```
i don't like that we specifically do this for POST (as i want to support configuring all request methods as cacheable). i think using the body in hash should not be an issue - its usually just empty. technically, GET requests can have a body as well, for example with elasticsearch. so its actually an improvement to also use the body for the hash.

tuupola replied:
Agree on this one too. However some might consider changing the way how keys are generated a BC break. See discussion in #12. Although if I read it correctly the consensus in the end was this would not be considered a BC break.

dbu replied:
i would not consider that a BC break. chances are you replace the cache on a new deployment anyways, and even if not, its just a cache.

sagikazarmark replied:
So what's the verdict? Do we change this to allow any methods?

dbu reviewed (approved):
my first reaction was "wtf, why?", then i read soap and think thats a reasonable usecase. and after some more thinking, i actually feel we should allow arbitrary methods to be cached if we allow to configure that at all.

dbu commented:
btw, caching proxies like varnish usually do a GET when seeing a HEAD request they have no cache for. then they store the response and return just the headers to honor the HEAD request. but i don't think we want to do that here. if people will do a GET anyways, they should not do a HEAD request first. they might be implementing their own cache, but then they should probably not use the cache-plugin.

sagikazarmark commented on spec/CachePluginSpec.php:
```diff
@@ -97,6 +97,49 @@ function it_doesnt_store_post_requests(CacheItemPoolInterface $pool, CacheItemIn
         $this->handleRequest($request, $next, function () {});
     }
 
+    function it_stores_post_requests_when_allowed(CacheItemPoolInterface $pool, CacheItemInterface $item, RequestInterface $request, ResponseInterface $response, StreamFactory $streamFactory, StreamInterface $stream)
```
Can you please format this according to PSR-1/2?

tuupola replied:
There were some other PSR-2 related problems too so I made a separate PR #25.

sagikazarmark commented on spec/CachePluginSpec.php:
```diff
@@ -97,6 +97,49 @@ function it_doesnt_store_post_requests(CacheItemPoolInterface $pool, CacheItemIn
+
+        $response->getStatusCode()->willReturn(200);
+        $response->getBody()->willReturn($stream);
+        $response->getHeader('Cache-Control')->willReturn(array())->shouldBeCalled();
```
Please use short array syntax (StyleCI is turned off for spec files because of the snake_case).

tuupola replied:
I was following the CS of the rest of the file. Long `array()` syntax is used elsewhere the spec file.

sagikazarmark commented on spec/CachePluginSpec.php:
```diff
@@ -97,6 +97,49 @@ function it_doesnt_store_post_requests(CacheItemPoolInterface $pool, CacheItemIn
+        $response->getHeader('Expires')->willReturn(array())->shouldBeCalled();
+        $response->getHeader('ETag')->willReturn(array())->shouldBeCalled();
+
+        $pool->getItem('e4311a9af932c603b400a54efab21b6d7dea7a90')->shouldBeCalled()->willReturn($item);
```
Please split this too

sagikazarmark commented:
Yeah, this one was a shocker for me at first too. 😄 

For the record: we should probably rethink our BDD(-misuse) strategy and rather use PHPUnit or something. Tests are almost unreadable with all the PSR-7 mocks.

tuupola committed:
PSR-2 and short array syntax

tuupola committed:
Allow any valid request method as defined by RFC7230

Request method can contain any visible USASCII character except
delimiters "(),/:;<=>?@[\]{}

https://tools.ietf.org/html/rfc7230#section-3.1.1
https://tools.ietf.org/html/rfc7230#section-3.2.6

tuupola committed:
Include request body in cache key calculation

tuupola committed:
Make StyleCI happy

tuupola committed:
PHP 5.4 does not have ::class

tuupola commented:
Updated the PR. It is now possible to enable caching for any arbitrary request method which conforms RFC7230. Also request body is always used when creating the cache key.

dbu commented on src/CachePlugin.php:
```diff
@@ -273,12 +272,19 @@ private function configureOptions(OptionsResolver $resolver)
+        $resolver->setAllowedTypes('methods', 'array');
         $resolver->setAllowedValues('hash_algo', hash_algos());
+        $resolver->setAllowedValues('methods', function ($value) {
+            $matches = preg_grep('/[^[:alnum:]!#$%&\'*\/+\-.^_`|~]/', $value);
```
can you please add a comment here with the RFC number that defines this regular expression? just for future reference and to make it clear we did not come up with a random expression here.

dbu replied:
i am horrible with regular expressions. is this forcing upper case? we should either force that or uppercase ourselves when processing the options. PSR-7 forces method to upper case. when you specify something lower case, it would never match.

the doc should explain that if you specify methods, you also need to specify GET if you still want those cached. and that you should do specific clients with different setup if you do soap and other stuff, to be sure only the right requests get cached...

tuupola replied:
It does not force uppercase. `[:alnum:]` is alias for `[0-9A-Za-z]`. [RFC2616](https://tools.ietf.org/html/rfc2616#section-5.1.1) says method is case-insensitive so technically both `get` and `GET` should be valid but just different methods.

I have no strong feelings whether we should force uppercase or not. I think most of the time people probably use uppercase methods.

The regexp basically says is request method has any other characters than these the method is invalid.

dbu commented:
cool, i like this. 

commits need to be squashed and we should have a documentation pull request (no big thing, just mention the feature in the right place). then its ready to merge imo

tuupola committed:
Mention RFC7230 which defines valid request methods

tuupola commented:
Documentation added. GitHub now has squash before merge feature which I think is better than doing squash + force push myself.

dbu commented:
> It does not force uppercase. |[:alnum:]| is alias for |[0-9A-Za-z]|.
> RFC2616 <https://tools.ietf.org/html/rfc2616#section-5.1.1> says method
> is case-insensitive so technically both |get| and |GET| should be valid
> but just different methods.
> 
> I have no strong feelings whether we should force uppercase or not. I
> think most of the time people probably use uppercase methods.

PSR-7 says: 

     * While HTTP method names are typically all uppercase characters, HTTP
     * method names are case-sensitive and thus implementations SHOULD NOT
     * modify the given string.

but for example guzzle psr-7 says:
https://github.com/guzzle/psr7/blob/5f17a79615bd36d52d3b3ce44075aec02605d8bf/src/Request.php#L92-L97

    public function withMethod($method)
    {
        $new = clone $this;
        $new->method = strtoupper($method);
        return $new;
    }

i am surprised by this, but feel like then we should not force upper case but just make sure all our examples use all caps and maybe mention that the option is case sensitive.

dbu reviewed (changes requested):
please change the CachePlugin constructor phpdoc to mention the new option for `method` (and say that its case sensitive)

tuupola committed:
Add new methods option to docblock

dbu commented on src/CachePlugin.php:
```diff
@@ -45,6 +45,7 @@
      *     @var int $cache_lifetime (seconds) To support serving a previous stale response when the server answers 304
      *              we have to store the cache for a longer time than the server originally says it is valid for.
      *              We store a cache item for $cache_lifetime + max age of the response.
+     *     @var array $methods case insensitive list of request methods which can be cached.
```
its case sensitive, not insensitive

tuupola replied:
Ah yes of course. Brain malfunction...

tuupola committed:
Request methods are case sensitive

dbu commented:
looks good to me.

what do the others think? can we merge?

Nyholm commented on src/CachePlugin.php:
```diff
@@ -225,7 +225,7 @@ private function getCacheControlDirective(ResponseInterface $response, $name)
     private function createCacheKey(RequestInterface $request)
     {
-        return hash($this->config['hash_algo'], $request->getMethod().' '.$request->getUri());
+        return hash($this->config['hash_algo'], $request->getMethod().' '.$request->getUri().$request->getBody());
```
What if the body is super large? Do we care?

Also, I suggest a space (or any other non-common delimiter) between uri and body.

tuupola replied:
Without space this is backwards compatible with previous versions if body is empty. Adding a space there will invalidate old caches, but it is of course ok with me if needed.

Nyholm replied:
Great. I did not think about that. How about doing something like: 

```php
$body = (string) $request->getBody()
if (!empty($body)) {
  $body = ' ' . $body;
}

return hash($this->config['hash_algo'], $request->getMethod().' '.$request->getUri().$body);
```

dbu replied:
i think we do not need to worry about keeping caches valid over deployments. but nyholms idea is good. as GET usually has empty body, it happens to keep old caches.

regarding big body: this method is only executed on methods that should be cached, right? so body will be usually empty, except for things like soap which hopefully won't be big file uploads. you would not want to cache a file upload, anyways.

Nyholm replied:
GET and HEAD has (EDIT: generally) no bodies. If you change the default behavior, then it is on you if you get performance issues.

tuupola replied:
I can add it. Why is the space needed BTW?

dbu replied:
GET can have body, for example when talking to elasticsearch: https://www.elastic.co/guide/en/elasticsearch/reference/current/search-request-body.html

Nyholm replied:
To make a difference between these two requests

```http
POST /foo
Host: www.example.com

bar
```

```http
POST /foob
Host: www.example.com

ar
```

tuupola replied:
I feel we could unconditionally add the space. It is only cache so it is not that big deal if old cache gets invalidated when plugin is updated. What do you think?

Nyholm replied:
Hm, sorry for the late response. 

I would prefer not invalidating cache when we can.. Im using Google Translate paid services and I know I will have to pay 100USD if my cache gets invalidated. 

Technically, this is okey. But we could be more nice to our users. =)

Nyholm commented on src/CachePlugin.php:
```diff
@@ -45,6 +45,7 @@
      *     @var int $cache_lifetime (seconds) To support serving a previous stale response when the server answers 304
      *              we have to store the cache for a longer time than the server originally says it is valid for.
      *              We store a cache item for $cache_lifetime + max age of the response.
+     *     @var array $methods case sensitive list of request methods which can be cached.
```
hm.. Yes, they are case sensitive but lower case will always fail..

`@var array $methods list of request methods which can be cached. Methods should always be upper case.`

tuupola replied:
Why it will fail?

dbu replied:
when you use the guzzle psr implementation, the method is forced to uppercase. but i pasted the fragment from psr-7 somewhere here that says that the getMethod of the request SHOULD preserve case. so depending on implementations, you could need lower case headers. i think its correct to allow lower case and not transform case.

Nyholm replied:
> i think its correct to allow lower case and not transform case.

But that would lead to unexpected behavior. I think we should do a case insensitive comparison or force input to be uppercase.

tuupola replied:
Cache plugin also [converts the request method to uppercase](https://github.com/php-http/cache-plugin/blob/master/src/CachePlugin.php#L65). While this is not technically correct it is probably ok for 99% of the use cases. I think the question is should we follow RFC or not.

dbu replied:
oh, missed that. ok, that would be a BC break if we stop doing this...

whatever, lets uppercase the method and force configuration to use uppercase. (thats better than magically uppercasing and confusing people that expected some specific behaviour). i think this will be correct in 99.999999% of the cases. would be surprised if anybody ever wants `GET` to be cached but not `get`...

Nyholm replied:
Agreed.

Nyholm commented on src/CachePlugin.php:
```diff
@@ -273,12 +273,20 @@ private function configureOptions(OptionsResolver $resolver)
         $resolver->setAllowedValues('hash_algo', hash_algos());
+        $resolver->setAllowedValues('methods', function ($value) {
+            /* Any VCHAR, except delimiters. RFC7230 sections 3.1.1 and 3.2.6 */
+            $matches = preg_grep('/[^[:alnum:]!#$%&\'*\/+\-.^_`|~]/', $value);
```
This looks really complicated. Why not use `[A-Z]+`. That will also make sure you use upper values. Or am I too specific?

tuupola replied:
Unless I have misunderstood the specs request method can also contain lowercase letters. Valid request methods are described in [RFC7230-3.1.1](https://tools.ietf.org/html/rfc7230#section-3.1.1) which states it is a token. Token is defined in [RFC7230-3.2.6](https://tools.ietf.org/html/rfc7230#section-3.2.6) which says any visible `USASCII` character except delimiters `"(),/:;<=>?@[\]{}`.

Nyholm replied:
Sorry, yes you are correct. (Reading the code and RFC more carefully...). I was wrong assuming that `!#$%&` etc should not be a part of the HTTP method. 

HTTP methods are generally uppercase but PSR-7 specifies that case should not change. I think we are correct by running `strtoupper()`. This implementation will, however, always fail when using a lower case method in the configuration. I suggest changing the regex to fail directly if a lower-case method are used. 

```php
$matches = preg_grep('/[^[A-Z0-9]!#$%&\'*\/+\-.^_`|~]/', $value);
```

dbu replied:
were do we strtoupper? i think we should leave the method unchanged, given how psr-7 is specified. as mentioned above, guzzle psr-7 request does not follow the SHOULD of the spec.

Nyholm reviewed (commented):
Great PR. Thanks. I had some questions.

Nyholm commented:
See https://github.com/php-http/cache-plugin/pull/24#discussion_r101534179 and the method doc block. I think we should do a case insensitive comparison.

dbu commented:
fine for me

tuupola committed:
Do not allow lowercase request methods

Technically RFC7230 allows lowercase request methods but consensus was
they might be too confusing for users.

tuupola committed:
Add delimiter between uri and body

Without delimiter the following requests would get the same cache key:

POST /foo
bar

POST /foob
ar

tuupola committed:
Remove unneeded use statement

dbu commented:
> I would prefer not invalidating cache when we can.. Im using Google Translate paid services and I know I will have to pay 100USD if my cache gets invalidated.

> Technically, this is okey. But we could be more nice to our users. =)

ok, i guess that is a good argument in favor of not concating ' '.$body if the body is empty (which it usually is for GET). sorry, can you please adjust that @tuupola ? i think then we are really ready. @sagikazarmark or do you see any issues that need to be adressed?

tuupola committed:
Maintain cache key bc with v1.2.0

tuupola committed:
Make StyleCI happy

tuupola committed:
Make StyleCI happy

tuupola commented:
Done. I will update the docs when I know this is frozen.

dbu commented:
thanks a lot! lets wrap up the doc PR as well. work for more flexibility will continue in #3

Nyholm commented:
Thank you @tuupola
